### ISSUE TITLE
determine BibTeX database locally using \bibliography and \addbibresource

### ISSUE THREAD

**Issue by timmli** [tmalsburg/helm-bibtex#112]
It seems that usually BibTeX database files are globally specified using `(setq bibtex-completion-bibliography '("/path/to/bibtex-file-1.bib" "/path/to/bibtex-file-2.bib"))`. Is it also possible to do this implicitely depending on the `\bibliography` and  `\addbibresource` commands of the LaTeX file (respectively its TeX-master) from which `ivy-bibtex` is invoked?

**Comment by tmalsburg** [comment]
This is currently not possible but it sounds like a useful feature for people who work with multiple bibtex-files.  (I have just one giant bibtex file.)  PRs welcome but if you plan to work on it, let's discuss the details first. Perhaps we can avoid some unnecessary detours.

**Comment by timmli** [comment]
This would be an extremely useful feature for me, since I usually work within collaborative LaTeX projects where there's always a local BibTeX file (also for better control of the BibTeX output when dealing with fixed templates). So, in general, I don't use my giant file directly, but copy entries into a local file. 

There's a workaround, namely `(setq bibtex-completion-bibliography '("./references.bib"))`, but this still forces my colleagues to conform to my emacs settings :( 

I know this feature from LatexTools for Sublime Text, and I saw that  [ebib](http://joostkremers.github.io/ebib/) supports this, too (yet unfortunately no fuzzy search on complete entries).

**Comment by jagrg** [comment]
@timmli You could experiment with a file local variable, and maybe use some hook to restore its default value.

`% -*- eval:(setq bibtex-completion-bibliography '("~/references.bib")) -*-`

**Comment by jagrg** [comment]
If you just need those files added to the list, you could try something like this:

`% -*- eval:(add-bib-file) -*-`

```elisp
(defun add-bib-file ()
  (save-excursion
    (when (re-search-forward "\\\\bibliography{" nil t)
      (let* ((match (thing-at-point 'word t))
	     (file (concat default-directory match ".bib")))
	(add-to-list 'bibtex-completion-bibliography file))))) ; t to append
````

**Comment by tmalsburg** [comment]
@jagrg, that's what I would do, too. But if I understand correctly, Timm's colleagues are working with Emacs, too, and Timm doesn't want to mess with their configuration.

Once solution would be to have a separate function `ivy-bibtex-guess-bibliography` (or `ivy-bibtex-local-bibliography`) that tries to find references to bibtex files in the current buffer, perhaps using code from ebib.  If it finds something, it would go on and call ivy-bibtex for searching temporarily overriding the default bibtex files using dynamic binding.

**Comment by jagrg** [comment]
That works too. We need two functions from ebib (`ebib--get-local-databases`, `ebib--ensure-extension`) and the binding added to helm/ivy-bibtex:

```
(let ((bibtex-completion-bibliography
       (if (eq major-mode 'latex-mode)
	   (ebib--get-local-databases)
	 bibtex-completion-bibliography)))
```

**Comment by tmalsburg** [comment]
Timm could you please try this:

``` elisp
(require 'ebib)

(defun ivy-bibtex-with-local-bibliography ()
	(interactive)
	(let ((bibtex-completion-bibliography
				 (if (eq major-mode 'latex-mode)
						 (ebib--get-local-databases)
					 bibtex-completion-bibliography)))
		(ivy-bibtex)))

```

And then `M-x ivy-bibtex-with-local-bibliography`.

**Comment by jagrg** [comment]
Very nice. BTW, the code also works in org-mode. Here's a simplified version that doesn't need ebib and is compatible with org-ref (cc @jkitchin).

```elisp
(defun bibtex-completion--get-local-databases ()
  "Return a list of .bib files associated with the current file."
  (save-excursion
    (goto-char (point-min))
    (cond
     ((re-search-forward "\\\\\\(?:no\\)*bibliography{\\(.*?\\)}" nil t)
      (mapcar (lambda (x)
     		(concat default-directory x ".bib"))
     	      (split-string (match-string-no-properties 1) ",[ ]*")))
     ((re-search-forward "^[ \t]*bibliography:\\([[:alnum:]].*\\)" nil t)
      (mapcar (lambda (x)
		(concat default-directory x))
	      (split-string (match-string-no-properties 1) ",[ ]*")))
     (t
      bibtex-completion-bibliography))))
```

```elisp
(defun helm-bibtex-with-local-bibliography ()
  "Search BibTeX entries with local bibliography."
  (interactive)
  (let ((bibtex-completion-bibliography
	 (if (or (eq major-mode 'latex-mode)
		 (eq major-mode 'org-mode))
	     (bibtex-completion--get-local-databases)
	   bibtex-completion-bibliography)))
    (helm-bibtex)))
```

**Comment by tmalsburg** [comment]
Thank you Jonathan, this is great.  Some comments:

I think this code should also be able to deal with references to bibtex files in biblatex format (`addbibresource`, see `ebib--get-local-databases`).

The function `ebib--get-local-databases` also uses `save-match-data` which seems like a good idea.

I'm not familiar with `default-directory`.  Is it necessarily the directory of the latex/org mode file?

I think your code always prepends `default-directory` to the path, but the path can also be absolute.  Your code also appends ".bib" unconditionally but Latex actually allows that you include the ".bib".  (See `ebib--ensure-extension`.)

`ebib--get-local-databases` uses `expand-file-name` to create a complete path if necessary.  Perhaps, we can just copy that. 

Thanks for looking into this.

**Comment by jagrg** [comment]
> I think this code should also be able to deal with references to bibtex files in biblatex format (addbibresource, see ebib--get-local-databases).

Done.

> The function ebib--get-local-databases also uses save-match-data which seems like a good idea.

Done.

> I'm not familiar with default-directory. Is it necessarily the directory of the latex/org mode file?

It seems to work fine without it.

> [...] Your code also appends ".bib" unconditionally but Latex actually allows that you include the ".bib". (See ebib--ensure-extension.)

Done.

> ebib--get-local-databases uses expand-file-name to create a complete path if necessary. Perhaps, we can just copy that.

I don't think this is necessary. WDYT?

```elisp
(defun bibtex-completion--get-local-databases ()
  "Return a list of .bib files associated with the current file."
  (save-excursion
    (save-match-data
      (goto-char (point-min))
      (cond
       ;; bibtex
       ((re-search-forward "\\\\\\(?:no\\)*bibliography{\\(.*?\\)}" nil t)
	(mapcar (lambda (fname)
		  (if (file-name-extension fname)
		      fname
		    (concat fname ".bib")))
		(split-string (match-string-no-properties 1) ",[ ]*")))
       ;; biblatex
       ((re-search-forward "\\\\addbibresource\\(\\[.*?\\]\\)?{\\(.*?\\)}" nil t)
	(mapcar (lambda (fname)
		  (if (file-name-extension fname)
		      fname
		    (concat fname ".bib")))
		(split-string (match-string-no-properties 2) ",[ ]*")))
       ;; org-ref
       ((re-search-forward "^[ \t]*bibliography:\\([[:alnum:]].*\\)" nil t)
	(mapcar (lambda (fname)
		  fname)
		(split-string (match-string-no-properties 1) ",[ ]*")))
       (t
	bibtex-completion-bibliography)))))
```

**Comment by timmli** [comment]
@tmalsburg, thanks for the ebib workaround! It seems to work flawlessly, even in non-master files (ebib then also checks the master file for bibliographies, which is desirable in larger projects).

@jagrg, I like your solution because it avoids to load ebib. But your function right now does not look into the master file when used in non-master files ...

**Comment by jagrg** [comment]
Here's a more thorough adaptation. I'm not too familiar with using a master file, so I'm not sure if this will work.

```elisp
(defun bibtex-completion--get-local-databases ()
  "Return a list of .bib files associated with the current file."
  (let ((texfile nil)
	(cb (current-buffer)))
    (when (and (boundp 'TeX-master)
	       (stringp TeX-master))
      (setq texfile (if (file-name-extension TeX-master)
			TeX-master
		      (concat TeX-master ".tex"))))
    (with-temp-buffer
      (if (and texfile (file-readable-p texfile))
	  (insert-file-contents texfile)
	(insert-buffer-substring cb))
      (save-match-data
	(goto-char (point-min))
	(cond
	 ;; bibtex
	 ((re-search-forward "\\\\\\(?:no\\)*bibliography{\\(.*?\\)}" nil t)
	  (mapcar (lambda (fname)
		    (if (file-name-extension fname)
			fname
		      (concat fname ".bib")))
		  (split-string (match-string-no-properties 1) ",[ ]*")))
	 ;; biblatex
	 ((re-search-forward "\\\\addbibresource\\(\\[.*?\\]\\)?{\\(.*?\\)}" nil t)
	  (mapcar (lambda (fname)
		    (if (file-name-extension fname)
			fname
		      (concat fname ".bib")))
		  (let ((option (match-string 1))
			(file (match-string-no-properties 2)))
		    (unless (and option (string-match-p "location=remote" option))
		      (split-string file ",[ ]*")))))
	 (t
	  bibtex-completion-bibliography))))))
```

**Comment by timmli** [comment]
@jagrg: Great, thanks! Your adaptation works as desired so far.

**Comment by tmalsburg** [comment]
I'm not familiar with master files either.  @jagrg, could you please make a PR?  I will then do some testing and merge if it works as planned.  Thank you!

**Comment by yilihong** [comment]
+1
this will be very useful.

**Comment by yilihong** [comment]
could we merge this with the master?

**Comment by tmalsburg** [comment]
Just merged #137 which adds this feature.   You can now do `helm-bibtex-with-local-bibliography` and `ivy-bibtex-with-local-bibliography`.  Thanks everyone for your contributions!
